### PR TITLE
Initial support for selecting the chunk size for parallel regions.

### LINKED ISSUE #2649: parallel schedule
The numba parellel with prange is awesome, but I often encounter for loops where the work needed to be done for each "i" in the loop is not the same. A scheduling option like the one in OpenMP would be a nice addition to the numba parallel feature.

For OpenMP this is descirbed as:

"dynamic: 
Use the internal work queue to give a chunk-sized block of loop iterations to each thread. When a thread is finished, it retrieves the next block of loop iterations from the top of the work queue. By default, the chunk size is 1. Be careful when using this scheduling type because of the extra overhead involved."

### PR DESCRIPTION
**Pull request by DrTodd13**
Resolves #2649 

Adding set_parallel_chunksize, get_parallel_chunksize, and parallel_chunksize (for use in "with" clauses).  The chunksize is stored as a thread-local variable and so keeps its value within a given thread until changed.  A value of 0 is the existing scheduling behavior which is the equivalent of OpenMP static schedule.  If a non-zero value is given, that is the chunk size, the number of iterations of the parfor that are bundled together for scheduling and is the equivalent of OpenMP dynamic scheduling with a specified chunk size.

I added a call to calculate the number of chunks we'll need and then I need to allocate that space in order to call the existing gufunc_scheduler do_scheduling routines.  Since that space is only known at runtime, you can't move the alloca for the schedule to the entry_block like every other variable now.  I added an option to keep the alloca where it is but this is a point to consider if we need to handle this some other way.

EDIT: I added some testing but for the raw functionality but there's still nothing that verifies that the chunks are of the right size or that we get the performance improvement we expect.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the PR @DrTodd13. Was wondering if you had any thoughts on what should happen in the nested scenario, for example, something like:
```python
@njit(parallel=True)
def bar(z):
  return z + 100

@njit(parallel=True)
def foo(x):
  a = np.zeros(1000,) # uses default chunking
  with set_parallel_chunksize(10):
    a[:] = x # uses chunking of size 10
    b = bar(a) # what happens here?! bar is nested parallel
    return b

foo(3)
```

**Comment by DrTodd13** [comment]
> Thanks for the PR @DrTodd13. Was wondering if you had any thoughts on what should happen in the nested scenario, for example, something like:

Is the call to bar() really nested parallel in this case?  You're not in a prange when you call it so you are initiating an outer parallel region inside bar.  So, given this code the way it is, this is the normal case I think where the chunksize gets set in the current thread and that thread calls bar and would use that chunksize for the parallel operation inside bar.

If you had the call to bar inside a prange so that the parallel region inside bar was a truly nested parallelism case then the question is more interesting.  All the outer prange workers are new workers so they'd all get the default chunksize of 0, right?  However, the internal chunking is also dependent on the thread count.  By default, isn't that going to be 0 or 1 for a prange worker which would mean that only one thread would work on all the chunks, however many there might be so the default chunksize of 0 would be appropriate.

So, the even more interesting case is where the programmer limits the threads in the outer parallel region and configures larger numbers of threads for the inner parallel regions.  In this case, the default would go back again to a static scheduling for those inner workers.  This isn't crystal clear in my head but instinct is telling me that the outer chunksize and inner chunksize would usually be uncorrelated, so getting the default static scheduling for the inner workers is okay and if you want different behavior you can use the same mechanisms to change it.

**Comment by stuartarchibald** [comment]
> > Thanks for the PR @DrTodd13. Was wondering if you had any thoughts on what should happen in the nested scenario, for example, something like:
> 
> Is the call to bar() really nested parallel in this case? You're not in a prange when you call it so you are initiating an outer parallel region inside bar. So, given this code the way it is, this is the normal case I think where the chunksize gets set in the current thread and that thread calls bar and would use that chunksize for the parallel operation inside bar.
> 

No, apologies, I forgot to put in a prange!

> If you had the call to bar inside a prange so that the parallel region inside bar was a truly nested parallelism case then the question is more interesting. All the outer prange workers are new workers so they'd all get the default chunksize of 0, right? However, the internal chunking is also dependent on the thread count. By default, isn't that going to be 0 or 1 for a prange worker which would mean that only one thread would work on all the chunks, however many there might be so the default chunksize of 0 would be appropriate.
> 
> So, the even more interesting case is where the programmer limits the threads in the outer parallel region and configures larger numbers of threads for the inner parallel regions. In this case, the default would go back again to a static scheduling for those inner workers. This isn't crystal clear in my head but instinct is telling me that the outer chunksize and inner chunksize would usually be uncorrelated, so getting the default static scheduling for the inner workers is okay and if you want different behavior you can use the same mechanisms to change it.

I'm inclined to agree, but perhaps it's worth seeing what OpenMP does? Also, given the chunk size is in TLS, I think there'll have to be some way of setting/unsetting the value internally in the threading implementation to handle cases where e.g. the outer prange main thread joins the threadpool and then subsequently also joins an inner prange threadpool, the value in the TLS slot would be carried. The machinery/design is present for this as its similar to the requirements and implementation of thread masking.

**Comment by DrTodd13** [comment]
> I'm inclined to agree, but perhaps it's worth seeing what OpenMP does? Also, given the chunk size is in TLS, I think there'll have to be some way of setting/unsetting the value internally in the threading implementation to handle cases where e.g. the outer prange main thread joins the threadpool and then subsequently also joins an inner prange threadpool, the value in the TLS slot would be carried. The machinery/design is present for this as its similar to the requirements and implementation of thread masking.

Yes, I wondered about this case but I'm not familiar with where that machinery is.  Could you point me in the right direction?  You are right that this has basically the same requirements as setting the number of threads.

**Comment by stuartarchibald** [comment]
> > I'm inclined to agree, but perhaps it's worth seeing what OpenMP does? Also, given the chunk size is in TLS, I think there'll have to be some way of setting/unsetting the value internally in the threading implementation to handle cases where e.g. the outer prange main thread joins the threadpool and then subsequently also joins an inner prange threadpool, the value in the TLS slot would be carried. The machinery/design is present for this as its similar to the requirements and implementation of thread masking.
> 
> Yes, I wondered about this case but I'm not familiar with where that machinery is. Could you point me in the right direction? You are right that this has basically the same requirements as setting the number of threads.

I'd probably take a look around here https://github.com/numba/numba/blob/master/numba/np/ufunc/parallel.py#L583-L627 to start with. I think this feature will need to work with `jit` disabled too, so the approach taken there for `get_num_threads` is probably along the right lines. In terms of carrying values across TLS slots, this is likely going to be backend specific, I'd start with OpenMP https://github.com/numba/numba/blob/master/numba/np/ufunc/omppool.cpp#L158 as it's probably easiest to reason about, and then:
* add something that locally captures the TLS value of the thread invoking the parallel region
* add something that sets the TLS chunksize to zero in the parallel region (so carried value for nested regions is zero)
* add something on exit from the parallel region that sets the TLS chunksize back to the captured value.

Whilst I think this could work, I'm inclined to give it a bit more thought to make sure it's going to be safe. Thoughts?

File: numba/np/ufunc/gufunc_scheduler.cpp [hunk 1]
```diff
@@ -11,6 +11,16 @@
...
 
+#ifdef _MSC_VER
+#define THREAD_LOCAL(ty) __declspec(thread) ty
+#else
+/* Non-standard C99 extension that's understood by gcc and clang */
+#define THREAD_LOCAL(ty) __thread ty
+#endif
+
+// Default 0 value means one evenly-sized chunk of work per worker thread.
+THREAD_LOCAL(uintp) parallel_chunksize = 0;
```
Comment (Hardcode84) [inline numba/np/ufunc/gufunc_scheduler.cpp:22]: Do Numba really support pre-C++11 compilers? In C++11 you can just use `thread_local`.

Also, you can make this var static or put it in anonymous namespace. And most of this file contents can be put into anonymous namespace.

Comment (DrTodd13) [reply numba/np/ufunc/gufunc_scheduler.cpp:22]: I will make it static.  In terms of the compiler, I did it this way because that is how it is done several places in Numba.  Maybe they should all change but I don't know why it is the way it is.  It could be legacy that could go away or there could still be a current reason for it.

Comment (stuartarchibald) [reply numba/np/ufunc/gufunc_scheduler.cpp:22]: It's because of historically having to support Python 2.7 and older compilers. If there's a cleaner way of doing this then that's a good change to make, but perhaps this PR should focus on functionality and then in a subsequent PR we can clean the numerous implementations doing the above?

Comment (Hardcode84) [reply numba/np/ufunc/gufunc_scheduler.cpp:22]: > perhaps this PR should focus on functionality and then in a subsequent PR we can clean the numerous implementations doing the above?

Ok, sounds good

**Comment by DrTodd13** [comment]
> * add something that locally captures the TLS value of the thread invoking the parallel region
> * add something that sets the TLS chunksize to zero in the parallel region (so carried value for nested regions is zero)
> * add something on exit from the parallel region that sets the TLS chunksize back to the captured value.

@stuartarchibald Done.  I also added some tests including one that makes sure that the full iteration space is processed (there was a bug with the original submission where that wasn't the case) and that all the threads running those iterations see their chunksize set to the default.  There was indeed a bug in this regard with the original submission where the invoking thread also ran some iterations and so didn't have its chunksize fixed.  I have resolved those issues so please re-review the PR.  Thanks!

**Comment by DrTodd13** [comment]
The test failure seems to be that TBB bug and not with this PR itself.

**Comment by stuartarchibald** [comment]
@DrTodd13 please could you resolve the conflicts so that a new build can be run. Thanks!

**Comment by DrTodd13** [comment]
> @DrTodd13 please could you resolve the conflicts so that a new build can be run. Thanks!

@stuartarchibald I resolved the conflicts and fixed some flake8 and strict prototype stuff and looks good now.

**Review by stuartarchibald — COMMENTED** [review]
@DrTodd13 Thanks for implementing the above suggestions. I've given the code a first review to provide some early feedback. Will now start looking at how the implementation behaves in practice by writing some more contrived tests, also want to look at the impact of setting chunksize in nested regions to zero, and investigate the usual potential issues arising in threading/forking.

File: numba/tests/test_parfors.py [hunk 2]
```diff
@@ -3739,6 +3740,78 @@ def find_maxima_3D_jit(args):
             find_maxima_3D_jit.py_func(args),
         )
 
+    @skip_parfors_unsupported
```
Comment (stuartarchibald) [inline numba/tests/test_parfors.py:3743]: Suggest all these tests go into a separate module `numba.tests.test_parallel_scheduling` or similar.

File: numba/__init__.py [hunk 3]
```diff
@@ -55,6 +55,11 @@
 # Initialize withcontexts
 import numba.core.withcontexts
 from numba.core.withcontexts import objmode_context as objmode
+from numba.core.withcontexts import parallel_chunksize
+
+# Re-export ParallelAccelerator runtime controls
+from numba.np.ufunc.parallel import (set_parallel_chunksize,
+                                     get_parallel_chunksize)
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/__init__.py:59]: Done.

File: numba/__init__.py [hunk 4]
```diff
@@ -80,6 +85,8 @@
     literal_unroll
     get_num_threads
     set_num_threads
+    set_parallel_chunksize
+    get_parallel_chunksize
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/__init__.py:85]: @stuartarchibald set_parallel_chunksize(0) is equivalent to OpenMP schedule(static).  set_parallel_chunksize(N>0) is equivalent to OpenMP schedule(dynamic,N).  I was taking inspiration from OpenMP when I was doing this.  I'm not going to argue on naming if you have a strong preference but I can tell you why I named it what I did.  Basically, I just wanted to associate these functions with the njit parallel option.  If we call them {g,s}et_schedule_chunksize() then we rely more on documentation to form the association that these are acting in conjunction with the parallel option.

I'll leave the decision up to you.

File: numba/core/withcontexts.py [hunk 5]
```diff
@@ -337,3 +338,65 @@ def _mutate_with_block_callee(blocks, blk_start, blk_end, inputs, outputs):
         outputs=outputs,
     )
 
+class _ParallelChunksize(WithContext):
+    is_callable = True
+
+    """A simple context-manager that tells the compiler to bypass the body
+    of the with-block.
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/core/withcontexts.py:408]: Good catch!

[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/core/withcontexts.py:408]: Done.

File: numba/core/cgutils.py [hunk 6]
```diff
@@ -359,7 +359,7 @@ def _setvalue(self, value):
     # __iter__ is derived by Python from __len__ and __getitem__
 
 
-def alloca_once(builder, ty, size=None, name='', zfill=False):
+def alloca_once(builder, ty, size=None, name='', zfill=False, entry_block=True):
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/core/cgutils.py:362]: Done.

File: numba/parfors/parfor_lowering.py [hunk 7]
```diff
@@ -1527,10 +1522,32 @@ def load_range(v):
...
+    get_sched_size = builder.module.get_or_insert_function(get_sched_size_fnty, name="get_sched_size")
+    num_divisions = builder.call(get_sched_size, [num_threads,
+                                                  context.get_constant(types.uintp, num_dim),
+                                                  dim_starts,
+                                                  dim_stops])
+    builder.call(set_chunksize, [zero])
+
+    multiplier = context.get_constant(types.uintp, num_dim * 2)
+    sched_size = builder.mul(num_divisions, multiplier)
+    sched = cgutils.alloca_once(builder, sched_type, size=sched_size, name="sched", entry_block=False)
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/parfors/parfor_lowering.py:1555]: Done.

File: docs/source/user/parallel.rst [hunk 8]
```diff
@@ -551,4 +551,90 @@ The report is split into the following sections:
...
+    set_parallel_chunksize(4)
+    func1()
+    func2()
+    func1()
+    set_parallel_chunksize(old_chunksize)
+
+Since this idiom of saving and restoring is so common, Numba provides the
+:func:`parallel_chunksize` with clause to simplify the idiom.  As shown below,
+this with clause can be invoked from both standard Python and within Numba
+jitted functions.
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply docs/source/user/parallel.rst:618]: @stuartarchibald What is your suggestion?  A change to the documentation to remind users that return is not allowed within a context managed block?

File: numba/np/ufunc/gufunc_scheduler.cpp [hunk 9]
```diff
@@ -68,8 +78,35 @@ class RangeActual {
...
+}
+
+extern "C" uintp get_sched_size(uintp num_threads, uintp num_dim, intp *starts, intp *ends) {
+    if (parallel_chunksize == 0) {
+        return num_threads;
+    }
+    RangeActual ra(num_dim, starts, ends);
+    uintp total_work_size = ra.total_size();
+    uintp num_divisions = total_work_size / parallel_chunksize;
+    return num_divisions < num_threads ? num_threads : num_divisions;
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/np/ufunc/gufunc_scheduler.cpp:107]: @stuartarchibald We already deviate from the OpenMP model because we have multi-dimensional parfors (and the chunksize has to apply to apply to the combination of those dimensions).  I'm not clear on this OpenMP point of how they could have nested parfors with different kinds of schedules potentially in combination with the collapse option.  Anyway,  I was inspired by OpenMP but not bound to it.  The chunksize you request is actually approximate.  As in the above code, we first turn the chunksize into the number of divisions of the work.  Do you agree with line 107?  If the array is 100 long and they have 4 threads and they specify a chunksize of 100 then do we use 4 threads or just 1?

If you had a 2D parfor and specified a chunksize of 5 then you are never going to get any chunk of work with exactly 5  items in it.  The system will try to make the average 5 but some will be larger and some smaller (4 or 6).

For your example, our system would currently see that 1 < 10/7 < 2 and so would need two divisions of the work. Then, it would do 2 equal divisions so you'd get 5 and 5.

File: numba/np/ufunc/parallel.py [hunk 10]
```diff
@@ -664,3 +664,55 @@ def impl():
 
 if _DYLD_WORKAROUND_SET and _DYLD_WORKAROUND_VAL:
     _launch_threads()
+
+
+def parfor_load_late(lib):
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/np/ufunc/parallel.py:669]: Done.  Integrated with _load_threading_functions and parfor_load_late is gone.

File: numba/core/withcontexts.py [hunk 11]
```diff
@@ -399,3 +400,68 @@ def _mutate_with_block_callee(blocks, blk_start, blk_end, inputs, outputs):
...
+        restore_state.append(ir.Assign(restore_spc_call, unused_var, loc))
+
+        blocks[blk_start].body = (blocks[blk_start].body[1:-1] + 
+                                  set_state + 
+                                  [blocks[blk_start].body[-1]])
+        blocks[blk_end].body = restore_state + blocks[blk_end].body
+        ir_utils.dprint_func_ir(func_ir, "After with changes", blocks=blocks)
+
+    def __call__(self, *args, **kwargs):
+        assert len(args) == 1
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/core/withcontexts.py:456]: Done.

File: numba/core/withcontexts.py [hunk 12]
```diff
@@ -337,3 +338,65 @@ def _mutate_with_block_callee(blocks, blk_start, blk_end, inputs, outputs):
...
+    """A simple context-manager that tells the compiler to bypass the body
+    of the with-block.
+    """
+    def mutate_with_body(self, func_ir, blocks, blk_start, blk_end,
+                         body_blocks, dispatcher_factory, extra):
+        ir_utils.dprint_func_ir(func_ir, "Before with changes", blocks=blocks)
+        assert extra is not None
+        args = extra["args"]
+        assert len(args) == 1
+        arg = args[0]
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/core/withcontexts.py:416]: Done.

File: numba/np/ufunc/parallel.py [hunk 13]
```diff
@@ -664,3 +664,55 @@ def impl():
...
+                                c_uint,
+                                c_uint,
+                                POINTER(c_int),
+                                POINTER(c_int))(lib.get_sched_size)
+
+
+def set_parallel_chunksize(n):
+    _launch_threads()
+    if not isinstance(n, (int, np.integer)):
+        raise TypeError("The parallel chunkize must be an integer")
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/np/ufunc/parallel.py:689]: Done.

File: numba/np/ufunc/parallel.py [hunk 14]
```diff
@@ -664,3 +664,55 @@ def impl():
...
+    global _get_parallel_chunksize
+    return _get_parallel_chunksize()
+
+
+@overload(set_parallel_chunksize)
+def ol_set_parallel_chunksize(n):
+    _launch_threads()
+    if not isinstance(n, types.Integer):
+        msg = "The parallel chunksize must be an integer"
+        raise errors.TypingError(msg)
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/np/ufunc/parallel.py:705]: Done.

File: numba/tests/test_parfors.py (hunk 2 above)
Comment (DrTodd13) [reply numba/tests/test_parfors.py:3743]: Hmm....How about just a different high-level test section in the existing test_parfors.py file since it is parfor related?

Comment (stuartarchibald) [reply numba/tests/test_parfors.py:3743]: Given the parfors only nature of this I agree with not moving it in hindsight. Having a new test class might be a good idea just purely for the purposes of grouping similar things.

File: docs/source/user/parallel.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply docs/source/user/parallel.rst:618]: I think a note about this would be useful as a reminder.

File: numba/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/__init__.py:85]: Thanks for explaining. Perhaps this should be `set_parfor_chunksize`? `guvectorize` and `vectorize` both support the `parallel` keyword but these functions have no effect on them. Alternatively perhaps this needs to be highlighted as a warning in the documentation.

File: numba/np/ufunc/gufunc_scheduler.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/np/ufunc/gufunc_scheduler.cpp:107]: I think this makes sense as it's bound to parfors, particularly RE 2d parfor. Also agree with l107, but think this example should go into the docs as I think it's quite likely a user would encounter it?

File: numba/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/__init__.py:85]: Added to documentation.

File: docs/source/user/parallel.rst [hunk 15]
```diff
@@ -551,4 +551,90 @@ The report is split into the following sections:
...
+        for i in prange(n):
+            ...
+        set_parallel_chunksize(old_chunksize)
+
+    old_chunksize = get_parallel_chunksize()
+    set_parallel_chunksize(4)
+    func1()
+    func2()
+    func1()
+    set_parallel_chunksize(old_chunksize)
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply docs/source/user/parallel.rst:613]: Done.

File: docs/source/user/parallel.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (DrTodd13) [reply docs/source/user/parallel.rst:618]: Done.

File: numba/np/ufunc/gufunc_scheduler.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/np/ufunc/gufunc_scheduler.cpp:107]: Done.

File: docs/source/user/parallel.rst [hunk 16]
```diff
@@ -551,4 +551,90 @@ The report is split into the following sections:
...
+    @njit(parallel=True)
+    def func2():
+        with parallel_chunksize(8):
+            for i in prange(n):
+                ...
+
+    with parallel_chunksize(4):
+        func1()
+        func2()
+        func1()
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply docs/source/user/parallel.rst:638]: Done but had to accept FreeVars in find_global_name in order for the with context code to recognize parallel_chunksize.

**Comment by stuartarchibald** [comment]
I think I may have a way to get the scheduler out of the parfors transforms, am struggling to debug issues with it, seems like there's some corruption going on somewhere. This `SIGABRT` due to double free (seems NRT dtor related):
```python
from numba import njit, prange, parallel_chunksize

@njit(parallel=True, debug=True)
def foo():
    with parallel_chunksize(2):
        acc = 0
        for i in prange(40):
            acc += i

print(foo())
```
do you see the same?

**Comment by DrTodd13** [comment]
> do you see the same?

@stuartarchibald No, I don't see this behavior.

**Comment by DrTodd13** [comment]
> > > do you see the same?
> > 
> > 
> > @stuartarchibald No, I don't see this behavior.
> 
> hmmm, it seems related to the number of threads/cores available, if you restrict to say 4, 3, 2? do you see it?

Yes, I see it non-deterministically at 2,3, or 4 threads.

**Comment by DrTodd13** [comment]
@stuartarchibald I know the problem.  Reduction arrays are still being allocated whose length is the number of threads.  However, we say in the gufunc descriptor that the schedule and the reduction arrays are the same length.  So, if the schedule gets bigger then you wind up writing to some memory past the end of the allocated reduction array.  There are two options as I see it:

1) We keep the reduction array equal to the number of threads but now the gufunc has to get it's thread number (from get_thread_id()?) and use that to select the right reduction array element.
2) We allocate the reduction array equal to size of the schedule.  That would require some significant code reorganization because schedule size is computed in create_gufunc...() but the reduction arrays are created before that in _lower_parfor_parallel().  Even worse, the code that initializes the reduction array at the beginning and does the final reduction at the end uses the thread count which is known at compile time.  There'd be significant complication to instead emit loops into the code equal to the length of the schedule.

I'm leaning to approach 1 but that relies on get_thread_id() which the documentation says should not be used outside of testing.  What do you think?

**Comment by stuartarchibald** [comment]
> @stuartarchibald I know the problem. Reduction arrays are still being allocated whose length is the number of threads. However, we say in the gufunc descriptor that the schedule and the reduction arrays are the same length. So, if the schedule gets bigger then you wind up writing to some memory past the end of the allocated reduction array.
> There are two options as I see it:
> 
>     1. We keep the reduction array equal to the number of threads but now the gufunc has to get it's thread number (from get_thread_id()?) and use that to select the right reduction array element.
> 
>     2. We allocate the reduction array equal to size of the schedule.  That would require some significant code reorganization because schedule size is computed in create_gufunc...() but the reduction arrays are created before that in _lower_parfor_parallel().  Even worse, the code that initializes the reduction array at the beginning and does the final reduction at the end uses the thread count which is known at compile time.  There'd be significant complication to instead emit loops into the code equal to the length of the schedule.
> 
> 
> I'm leaning to approach 1 but that relies on get_thread_id() which the documentation says should not be used outside of testing. What do you think?

@DrTodd13 Thanks for taking a look at this, it's great you've discovered the root cause. From your description and the changes made in the patch, this indeed sounds likely.

**RE: approach 1:**

Thoughts about the use of `get_thread_id`...  It was written for use in testing thread masking whereby all that mattered was that `n` threads were active in a region with mask `n`. The value the `get_thread_id` function returns is threading layer dependent:

 * Workqueue returns `(int)pthread_self()`. (NOTE TO SELF: this should be fixed to `size_t`).
 * OpenMP returns `omp_get_thread_num()`.
 * TBB returns `tbb::task_arena::current_thread_index()`.

I think all of these present a degree of "uniqueness" in that their IDs will not collide at any given time within the execution of a process. In practice, because of the way these threadpools are launched and run, I don't _think_ there's an issue with recycling of IDs, but I'm also reasonably convinced that this doesn't matter for what you propose so long as throughout the reduction process all IDs are unique and the set of IDs is constant. What I am concerned about though is dynamic scheduling of new threads and/or work steal i.e. a non-constant set of IDs during the reduction process. Say for example TBB starts work with IDs `[0, 1]`, then decides that some other thing with thread IDs `[2, 3]` is more pressing, could ID `1` move there to make the work have `[0,]` and other work have `[1, 2, 3]` . What happens if the load needs another rebalance and it decides to move ID `3` back to `work`, then you have `[0, 3]`, `[1, 2]`. I've no idea if this sort of thing happens in practice as I'm not familiar with how TBB decides how to schedule and move threads about, but it is something to consider.

Finally, these thread IDs are neither contiguous in value or guaranteed to start at 0, e.g. `pthread_self` from `workqueue` will return something unique for the process at that time but could be any pretty much any value. I think this means writes would have to go through a dictionary mapping `thread_id -> slot in reduction array`? This also assuming that the thread IDs working on a reduction are constant for the duration of the reduction (as noted previously).

**RE: approach 2:**

This sounds complicated! I believed we've previously run into trouble before with baked in sizes based on compile time thread counts (contrived and slightly unusual example in: https://github.com/numba/numba/issues/5890#issuecomment-661848040). It does also sound like a lot of restructuring would be required to get this to work as the code is already part lowered by the time the schedule is created. However, I think that this may ultimately be safest and produce more efficient code as there would be no dynamic indirection mapping threads to slots needed?

----------------
In light of all this, I'm concerned that there's potential for this change to break in production code as it seems like something fundamental to parfors (reductions computations) will need significant adjustment with no obvious way to isolate it.

I suggest:
* The `chunksize` API presented here goes under `numba.experimental` until we're more convinced of it, this also allows us to go in later and change the underlying behaviour to a lower level such that scheduling is adjusted at the threadpool level instead of in the schedule. This would potentially provide benefit in the `vectorize`/`guvectorize` cases too.
* As the above API location change doesn't cover the potentially large changes needed to parfors fundamentals we should discuss this aspect further.
* This all needs significant testing, particularly with variations in thread pool size and supplied chunk size, along with the expected API/use case tests.

Thoughts?

**Comment by Hardcode84** [comment]
I think both TBB and OpenMP layers guarantee `get_thread_id` to be in `[0:max_concurrency)` (where `max_concurrency` = value we passed in launch_threads) and implementing similar behavior for workqueue should be straightforward:
* Add `int thread_index;`  to `Queue` struct
* Assign them sequentially in `launch_threads`
* Copy val from queue to thread_local var in`thread_worker`

**Comment by stuartarchibald** [comment]
> I think both TBB and OpenMP layers guarantee `get_thread_id` to be in `[0:max_concurrency)` (where `max_concurrency` = value we passed in `launch_threads`) and implementing similar behavior for workqueue should be straightforward:
> 
>     * Add `int thread_index;`  to `Queue` struct
> 
>     * Assign them sequentially in `launch_threads`
> 
>     * Copy val from queue to thread_local var in`thread_worker`

@Hardcode84 for confirming RE TBB `thread_id`, I think what you could describe could potentially work for `workqueue`. Any thoughts on TBB moving threads about?

**Comment by Hardcode84** [comment]
I didn't quite get the question. `current_thread_index` is guaranteed to be between 0 and `num_threads` (value passed to `task_arena` ctor), but exact scheduling inside this range is unpredictable and it is entirely possible for all chunks to be executed on single thread for example and in this case they all will get same id

Also, its worth mentioning that `get_thread_id` only makes sense when called from inside parallel constructs.
`omp_get_thread_num` always returns 0 when called outside parallel region and `task_arena::current_thread_index` always return -1 when called outside task/parallel_for body.

**Comment by DrTodd13** [comment]
I am fine with this going in experimental.  I can try to create a version that uses get_thread_id() assuming that all thread implementations guarantee the thread ids go from [0,max_configured_threads).

**Comment by stuartarchibald** [comment]
Jotting down some outcomes from the Numba open development meeting today:

1. Make `_get_thread_id` "work" for `workqueue` (@stuartarchibald)
2. Publish the way to get the scheduler out of a `prange` block (@stuartarchibald)
3. Update `parfors` to reduce into the `max_configured_threads` sized array still, but to index subarrays using calls staged to `_get_thread_id`. (@DrTodd13)
4. Remove `_get_thread_id` from any user side docs, it is good for uses such as those in this PR/described above but in general doesn't have well enough defined semantics for use in user code cf. `omp_get_thread_num`. (TBD)
5. Write a lot of tests. (@stuartarchibald @DrTodd13).

**Comment by stuartarchibald** [comment]
@DrTodd13 I've got a rough cut of 1. and 2. above in https://github.com/numba/numba/pull/6674, patch 35d2b10. `_get_thread_id` should work ok now for the purposes needed here. The schedule fetching is caveat laden (and definitely not public!) but essential for testing, so we'll just have to be aware of the limitations of where it can be used (basically, only in a parallel region). I've been debugging it with something like:
```python
from numba import njit, prange, set_parallel_chunksize
from numba.np.ufunc.parallel import _get_thread_id
from numba.np.ufunc.parallel import _get_schedule
import numpy as np

for psz in range(20):
    for cs in range(psz + 1):
        @njit(parallel=True, debug=True)
        def foo():
            set_parallel_chunksize(cs)
            main_thread = _get_thread_id()
            sched = []
            for i in prange(psz):
                if _get_thread_id() == main_thread:
                    sched.append(_get_schedule())
            return sched[0] if sched else np.empty((0,0), 'uint64')
        print(f'loop = {psz}, chunk = {cs}'.center(80, '-'))
        print(foo())
```
the above is making the assumption that, for execution backends that incorporate the calling thread, that `main_thread` becomes part of the threadpool and takes part in the work. Another thing I've been doing is allocating an accumulating container the size of the thread mask and then using `_get_thread_id` to do safe writes of the schedule into that. Essentially, this sort of thing:
```python
from numba import njit, prange, set_parallel_chunksize, get_num_threads
from numba.np.ufunc.parallel import _get_thread_id
from numba.np.ufunc.parallel import _get_schedule
import numpy as np

for psz in range(20):
    for cs in range(psz + 1):
        @njit(parallel=True, debug=True)
        def foo():
            set_parallel_chunksize(cs)
            main_thread = _get_thread_id()
            sched = []
            for x in range(get_num_threads()):
                empty = []
                empty.append(np.empty((0,0), 'uint64'))
                empty.pop()
                sched.append(empty)
            for i in prange(psz):
                sched[_get_thread_id()].append(_get_schedule())
            for x in range(get_num_threads()):
                print("THREAD INDEX", x)
                print(sched[x])
        print(f'loop = {psz}, chunk = {cs}'.center(80, '-'))
```

**Comment by DrTodd13** [comment]
@stuartarchibald @sklam Okay, I found the problem with 2D+ reductions.  There is one problem that I wasn't able to root cause to my satisfaction and would probably be better if one of you looked into it.  In https://github.com/numba/numba/pull/6025/commits/42c01213a0a5d66b085ae5b3415225a30f6a0d47,  I added something to unify things of type Function with things of type ExternalFunctionPointer.  At some point, a getattr of numba.np.ufunc.parallel._get_thread_id() is resolved to ExternalFunctionPointer but somewhere else the lhs of that assignment was typed as Function.  This failed to unify but I couldn't find where the lhs was being typed that way and I'm not sure which type is correct.  I do know that if you unify the types as Function then it doesn't work but it does work if you unify them as ExternalFunctionPointer.

**Comment by songololo** [comment]
Just thinking out loud, but wouldn't this pull request also make some sort of progress bar possible?

i.e. would it be possible to track the number of completed chunks vs. total chunks so that every n-th iteration can invoke an obj-mode callback to something like `tqdm`?

If this would not be possible, is there a way to accommodate this sort of logic without overly complicating the gist of this change?

Thanks.

See #4267 and https://github.com/krishnanlab/PecanPy/issues/5
